File: online judge.kaner murillo/python/673.py
#https://onlinejudge.org/external/6/673.pdf
#673

from sys import stdin 
from collections import deque

casos = int(stdin.readline().strip())

for i in range(casos):

    cadena = stdin.readline().rstrip('\n')
    pila = deque()

    n = 0
    ans = True

    while n<len(cadena) and ans == True:

        if(cadena[n] == "[" or cadena[n] == "("):
            pila.append(cadena[n])
        
        elif (len(pila)!=0 and pila[-1] == "[" and cadena[n] == "]"):
            pila.pop()
        
        elif (len(pila)!=0 and pila[-1] == "(" and cadena[n] == ")"):
            pila.pop()

        else:
            ans = False

        n+=1
    
    if(len(pila)!=0):
        ans = False
    
    if ans == True:
        print("Yes")
    
    else:
        print("No")

File: online judge.kaner murillo/c++/673.cpp
//https://onlinejudge.org/external/6/673.pdf
//673

#include<cstdio>
#include<iostream>
#include<stack>
#include<string>

using namespace std;

int main(){

    int casos, i;
    bool ans;
    string cadena, letra;


    scanf("%i", &casos);
    cin.ignore();

    while(casos--){

        ans = true;
        stack<string>pila;

        getline(cin,cadena);

        i = 0;

        while(i<cadena.length() && ans == true){
            
            letra = cadena[i];

            if(letra == "[" || letra == "("){
                pila.push(letra);
            }

            else if(letra == "]" || letra == ")"){

                if(!pila.empty() && pila.top() == "[" && letra == "]"){
                    pila.pop();
                }

                else if(!pila.empty() && pila.top() == "(" && letra == ")"){
                    pila.pop();
                }

                else{
                    ans = false;
                }
            }
            i+=1;
        }

        if(!pila.empty()){
            ans = false;
        }

        if(ans == true){
            printf("Yes\n");
        }

        else{
            printf("No\n");
        }
    }
    return 0;
}